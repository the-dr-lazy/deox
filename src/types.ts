import { ActionCreator } from './create-action-creator'
import { Action, AnyAction } from './action'

type Primitive = undefined | null | boolean | string | number | Function

export interface DeepImmutableArray<T>
  extends ReadonlyArray<DeepImmutable<T>> {}
export interface DeepImmutableMap<K, V>
  extends ReadonlyMap<DeepImmutable<K>, DeepImmutable<V>> {}
export type DeepImmutableObject<T> = {
  readonly [K in keyof T]: DeepImmutable<T[K]>
}

export type Immutable<T> = T extends Primitive
  ? T
  : T extends Array<infer U>
  ? ReadonlyArray<U>
  : T extends Map<infer K, infer V>
  ? ReadonlyMap<K, V>
  : Readonly<T>

export type DeepImmutable<T> = T extends Primitive
  ? T
  : T extends Array<infer U>
  ? DeepImmutableArray<U>
  : T extends Map<infer K, infer V>
  ? DeepImmutableMap<K, V>
  : DeepImmutableObject<T>

export type Reducer<State, Actions> = (
  prevState: DeepImmutable<State>,
  action: Actions
) => DeepImmutable<State> | State

export type ActionType<
  T extends ActionCreator<AnyAction> | Reducer<any, Action<any>>
> = T extends ActionCreator<AnyAction>
  ? ReturnType<T>
  : (T extends Reducer<any, infer U> ? U : never)
