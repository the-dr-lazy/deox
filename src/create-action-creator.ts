import { createAction, Action, AnyAction } from './create-action'

export type ExactActionCreator<TType extends string, TCallable extends <_T>(...args: any[]) => Action<TType>> = TCallable & {
  type: TType extends AnyAction ? TType['type'] : TType
  toString(): TType extends AnyAction ? TType['type'] : TType
}

export type ActionCreator<T extends AnyAction | string> = {
  (...args: any[]): T extends string ? Action<T> : T
  type: T extends AnyAction ? T['type'] : T
  toString(): T extends AnyAction ? T['type'] : T
}

export type Executor<TType extends string, TCallable extends <_T>(...args: any[]) => Action<TType>> = (
  resolve: <Payload = undefined, Meta = undefined>(
    payload?: Payload,
    meta?: Meta
  ) => Action<TType, Payload, Meta>
) => TCallable;

/**
 * Flux standard action creator factory
 * @example
 * createActionCreator('[Todo] truncate');
 * @example
 * createActionCreator(
 *   '[Todo] fetch rejected',
 *   resolve => (error: Error) => resolve(error)
 * );
 * @example
 * createActionCreator(
 *   '[Todo] fetch rejected',
 *   resolve => (error: Error, meta?: { status: number }) => resolve(error, meta)
 * )
 * @example
 * createActionCreator(
 *   '[Todo] add',
 *   resolve => (name: string, completed = false) => resolve({ name, completed })
 * )
 * @example
 * createActionCreator(
 *   '[Todo] add',
 *   resolve => (name: string, completed = false) => resolve({ name, completed }, 'Meta data of all todos')
 * )
 *
 */
export function createActionCreator<TType extends string>(type: TType): ExactActionCreator<TType, () => Action<TType>>;
export function createActionCreator<
  TType extends string,
  TCallable extends <_T>(...args: any[]) => Action<TType>
>(type: TType, executor: Executor<TType, TCallable>): ExactActionCreator<TType, TCallable>;
export function createActionCreator<
  TType extends string,
  TCallable extends <_T>(...args: any[]) => Action<TType>
>(type: TType, executor: Executor<TType, TCallable> = resolve => (() => resolve()) as TCallable) {
  const callable = executor((payload, meta) =>
    createAction(type, payload!, meta!)
  )

  return Object.assign(callable, {
    type,
    toString() {
      return type
    },
  })
}
