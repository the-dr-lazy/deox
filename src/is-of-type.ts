import { AnyAction } from './create-action'
import { ActionCreator } from './create-action-creator'
import { getType } from './get-type'
import { castArray } from './utils'
import { ExtractAction } from './types'

/**
 * Non-curried function
 * @example
 * const increment = createActionCreator('[Counter] increment')
 * const decrement = createActionCreator('[Counter] decrement')
 * const reset = createActionCreator(
 *   '[Counter] reset',
 *   resolve => (value: number) => resolve(value)
 * )
 * isOfType(increment.type, increment()) //=> true
 * @example
 * isOfType([increment.type, decrement.type], increment()) //=> true
 * @example
 * isOfType(decrement(), increment()) //=> false
 */
export function isOfType<
  TSource extends string | AnyAction | ActionCreator<AnyAction>,
  TAction extends AnyAction
>(
  type: TSource | TSource[],
  action: TAction
): action is ExtractAction<TSource, TAction>

/**
 * Curried function
 * @example
 * const increment = createActionCreator('[Counter] increment')
 * const decrement = createActionCreator('[Counter] decrement')
 * isOfType(increment.type)(increment()) //=> true
 * @example
 * isOfType([increment.type, decrement.type])(increment()) //=> true
 * @example
 * isOfType(decrement())(increment()) //=> false
 */
export function isOfType<
  TSource extends string | AnyAction | ActionCreator<AnyAction>
>(
  type: TSource | TSource[]
): <TAction extends AnyAction>(
  action: TAction
) => action is ExtractAction<TSource, TAction>

/**
 * Action type assertion helper
 * @description Check if action type property is equal given action(s) type property or action creator(s) type property or action type(s).
 * @example
 * const increment = createActionCreator('[Counter] increment')
 * const decrement = createActionCreator('[Counter] decrement')
 * const reset = createActionCreator(
 *   '[Counter] reset',
 *   resolve => (value: number) => resolve(value)
 * )
 * isOfType([decrement(), reset(0)], increment()) //=> false
 * @example
 * isOfType(reset, increment()) //=> false
 * @example
 * isOfType([reset, increment], increment()) //=> true
 * @example
 * isOfType([increment.type, decrement(), reset], increment()) //=> true
 */
export function isOfType<
  TSource extends string | AnyAction | ActionCreator<AnyAction>,
  TAction extends AnyAction
>(keys: TSource | TSource[], action?: TAction) {
  const types = castArray<string | AnyAction | ActionCreator<AnyAction>>(
    keys
  ).map(key => (typeof key === 'string' ? key : getType(key)))

  const assertFn = (action: TAction) => types.includes(action.type)

  // 1 arg case => return curried version
  if (action === undefined) {
    return assertFn
  }

  // 2 args case => invoke assertFn and return the result
  return assertFn(action)
}
