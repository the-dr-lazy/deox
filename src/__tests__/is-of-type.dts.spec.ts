import { isOfType } from '../is-of-type'
import { createActionCreator } from '../create-action-creator'
import { ActionType } from '../types'

// @dts-jest:group isOfType

const increment = createActionCreator('[Counter] increment')
const decrement = createActionCreator('[Counter] decrement')
const reset = createActionCreator(
  '[Counter] reset',
  resolve => (value: number) => resolve(value)
)
const actions = [increment(), decrement(), reset(0)]

declare const action:
  | ActionType<typeof increment>
  | ActionType<typeof decrement>
  | ActionType<typeof reset>

{
  // @dts-jest:pass:snap
  actions.filter(isOfType(increment.type))

  // @dts-jest:pass:snap
  actions.filter(isOfType([increment.type, decrement.type]))

  // @dts-jest:pass:snap
  actions.filter(isOfType(decrement()))

  // @dts-jest:pass:snap
  actions.filter(isOfType([decrement(), reset(0)]))

  // @dts-jest:pass:snap
  actions.filter(isOfType(reset))

  // @dts-jest:pass:snap
  actions.filter(isOfType([reset, increment]))

  // @dts-jest:pass:snap
  actions.filter(isOfType([increment.type, decrement(), reset]))
}

{
  if (isOfType(increment.type, action)) {
    // @dts-jest:pass:snap
    action // tslint:disable-line:no-unused-expression
  }

  if (isOfType([increment.type, decrement.type], action)) {
    // @dts-jest:pass:snap
    action // tslint:disable-line:no-unused-expression
  }

  if (isOfType(decrement())(action)) {
    // @dts-jest:pass:snap
    action // tslint:disable-line:no-unused-expression
  }

  if (isOfType([decrement(), reset(0)])(action)) {
    // @dts-jest:pass:snap
    action // tslint:disable-line:no-unused-expression
  }

  if (isOfType(reset)(action)) {
    // @dts-jest:pass:snap
    action // tslint:disable-line:no-unused-expression
  }

  if (isOfType([reset, increment], action)) {
    // @dts-jest:pass:snap
    action // tslint:disable-line:no-unused-expression
  }

  if (isOfType([increment.type, decrement(), reset])(action)) {
    // @dts-jest:pass:snap
    action // tslint:disable-line:no-unused-expression
  }
}
