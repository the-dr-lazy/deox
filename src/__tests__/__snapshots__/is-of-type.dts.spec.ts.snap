// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`isOfType action (type) should match snapshot 1`] = `"{ type: \\"[Counter] increment\\"; }"`;

exports[`isOfType action (type) should match snapshot 2`] = `"{ type: \\"[Counter] increment\\"; } | { type: \\"[Counter] decrement\\"; }"`;

exports[`isOfType action (type) should match snapshot 3`] = `"{ type: \\"[Counter] decrement\\"; }"`;

exports[`isOfType action (type) should match snapshot 4`] = `"{ type: \\"[Counter] decrement\\"; } | { type: \\"[Counter] reset\\"; payload: number; }"`;

exports[`isOfType action (type) should match snapshot 5`] = `"{ type: \\"[Counter] reset\\"; payload: number; }"`;

exports[`isOfType action (type) should match snapshot 6`] = `"{ type: \\"[Counter] increment\\"; } | { type: \\"[Counter] reset\\"; payload: number; }"`;

exports[`isOfType action (type) should match snapshot 7`] = `"{ type: \\"[Counter] increment\\"; } | { type: \\"[Counter] decrement\\"; } | { type: \\"[Counter] reset\\"; payload: number; }"`;

exports[`isOfType actions.filter(isOfType([decrement(), reset(0)])) (type) should match snapshot 1`] = `"({ type: \\"[Counter] decrement\\"; } | { type: \\"[Counter] reset\\"; payload: number; })[]"`;

exports[`isOfType actions.filter(isOfType([increment.type, decrement(), reset])) (type) should match snapshot 1`] = `"({ type: \\"[Counter] increment\\"; } | { type: \\"[Counter] decrement\\"; } | { type: \\"[Counter] reset\\"; payload: number; })[]"`;

exports[`isOfType actions.filter(isOfType([increment.type, decrement.type])) (type) should match snapshot 1`] = `"({ type: \\"[Counter] increment\\"; } | { type: \\"[Counter] decrement\\"; })[]"`;

exports[`isOfType actions.filter(isOfType([reset, increment])) (type) should match snapshot 1`] = `"({ type: \\"[Counter] increment\\"; } | { type: \\"[Counter] reset\\"; payload: number; })[]"`;

exports[`isOfType actions.filter(isOfType(decrement())) (type) should match snapshot 1`] = `"{ type: \\"[Counter] decrement\\"; }[]"`;

exports[`isOfType actions.filter(isOfType(increment.type)) (type) should match snapshot 1`] = `"{ type: \\"[Counter] increment\\"; }[]"`;

exports[`isOfType actions.filter(isOfType(reset)) (type) should match snapshot 1`] = `"{ type: \\"[Counter] reset\\"; payload: number; }[]"`;